spec/models/wunderground_history_spec.rb

```diff
@@ -2,37 +2,88 @@
-      wh.time = Time.zone.parse('January 1, 3000 00:00:00 -04:00')
-      wh.response = wunderground.history_for(wh.time, 'knyc')
-      expect(wh).to be_premature
-    end
+    describe "validate!" do
+      context "when history is empty" do
+        let(:empty_response) { wunderground.history_for(out_of_range_time, 'knyc') }
+        let(:empty_history) do
+          build(:wunderground_history, {
+            time: out_of_range_time,
+            response: empty_response
+          })
+        end
 
-    it "returns false when data was available at time of creation" do
-      wh.time = Time.zone.parse('January 1, 2000 00:00:00 -04:00')
-      wh.response = wunderground.history_for(wh.time, 'knyc')
-      expect(wh).to_not be_premature
+        it "adds 'response is empty' error" do
+          invalid_time_message = {response: ["response is empty"]}
+          empty_history.validate!
+
+          expect(empty_history.errors).to have(1).errors_on(:response)
+          expect(empty_history.errors.messages).to eq(invalid_time_message)
+        end
+      end
+
+      it "adds error when rate limited" do
+        subject.time = time
+        subject.response = response
+        subject.validate!
+        expect(subject).to be_rate_limited # only re-record while rate-limited
+      end
+
+      it "adds error when location is invalid" do
+        subject.time = Time.zone.parse('January 1, 2000 at 12am')
+        subject.response = wunderground.history_for(subject.time, '1001101')
+        subject.validate!
+        expect(subject.response_error_type).to eq 'querynotfound'
+        expect(subject.errors).to have(1).errors_on(:response)
+      end
     end
-  end
 
-  describe "rate_limited?" do
-    before(:each) do
+    describe "missing_target_hour?" do
+      it "returns true when no data yet available from wunderground" do
+        subject.time = Time.zone.parse('January 1, 3000 at 12am')
+        subject.response = wunderground.history_for(subject.time, 'knyc')
+        expect(subject).to be_missing_target_hour
+      end
+
+      it "returns false when data was available at time of creation" do
+        subject.time = Time.zone.parse('January 1, 2000 at 12am')
+        subject.response = wunderground.history_for(subject.time, 'knyc')
+        expect(subject).to_not be_missing_target_hour
+      end
     end
 
-    it "returns true when rate limited" do
-      wh.time = Time.zone.parse('January 1, 2000 00:00:00 -04:00')
-      wh.response = wunderground.history_for(wh.time, 'knyc')
-      expect(wh).to be_rate_limited
+    describe "empty?" do
+      it "returns true when no data yet available from wunderground" do
+        subject.time = Time.zone.parse('January 1, 3000 at 12am')
+        subject.response = wunderground.history_for(subject.time, 'knyc')
+        expect(subject).to be_empty
+      end
+
+      it "returns false when data was available at time of creation" do
+        subject.time = Time.zone.parse('January 1, 2000 at 12am')
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.